### PR TITLE
IGNITE-13674  Documentation page added for Native Persistence Defragmentation feature

### PR DESCRIPTION
**Pull request by sergey-chugunov-1985**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+= Persistence Defragmentation
+
+== Overview
+
+As memory management mechanism of Apache Ignite can only create or reuse pages for user data but never frees them files where Ignite persists data can only grow and never shrinks.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:19]:
```suggestion
Apache Ignite memory management mechanism can only create or reuse pages for user data, but it never frees them. So files, where Ignite persists data, can only grow, but never shrink.
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+In most use cases it doesn't cause any problems as once created page can be reused multiple times. However in certain cases it is possible that cache contains very little data but occupies large chunks of disk space because a lot of data was removed from the cache.
+
+Defragmentation is aimed to enable user to shrink data files and claim back disk space.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:23]:
```suggestion
Defragmentation enables user to shrink data files and claim back disk space.
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+[NOTE]
+====
+Defragmentation can only be used with historical rebalance enabled (link to historical rebalance page). If historical rebalance is disabled server node always triggers full rebalance after restart throwing away defragmented partition. Full set of data is transferred to the node from other nodes over network, depending of size of data set it may require a lot of time and may slow down the whole cluster as network capacity is important to fulfill user requests.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:27]:
```suggestion
Defragmentation can only be used with historical rebalance enabled (link to historical rebalance page). If historical rebalancing is disabled, the server node always triggers full rebalance after the restart, which would throw away the defragmented partition. A full set of data is transferred to the node from other nodes over a network. Depending on the dataset’s size, transferring may require significant time and slow down the whole cluster as network capacity is essential to fulfill user requests.
```

Comment (Nikita-tech-writer) [reply docs/_docs/persistence/native-persistence-defragmentation.adoc:27]: @sergey-chugunov-1985 Please add a link to the **Historical rebalance** page.

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+====
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:28]:
```suggestion

```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+== Usage: how and when
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:30]:
```suggestion
== Performing Defragmentation
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+Defragmentation is costly operation in terms of disk IO so to avoid slowing down user operations it cannot be executed on regular node joined to the cluster. To execute defragmentation user needs to request it first on a particular node or set of nodes and than restart these nodes.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:32]:
```suggestion
Defragmentation is a costly operation in terms of disk IO. To avoid slowing down user operations, do not execute it on a regular node joined to the cluster. To perform defragmentation, you need to request it first on a particular node or set of nodes and then restart them.
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+To request defragmentation use the following command: *control.(sh|bat) --defragmentation schedule --nodes <consistentIds> [--caches <cacheNames>]*.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:34]:
```suggestion
=== Starting Defragmentation

To request defragmentation use the following command: *control.(sh|bat) --defragmentation schedule --nodes <consistentIds> [--caches <cacheNames>]*.
```

Comment (Nikita-tech-writer) [reply docs/_docs/persistence/native-persistence-defragmentation.adoc:34]: Please use a code block for the command here

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+After restart node with requested defragmentation will enter special mode called maintenance mode. Node in maintenance doesn't join the rest of the cluster but stays isolated until defragmentation is completed (or cancelled by explicit user request). After that user has to restart the node one more time: it will exit maintenance mode and returns back to normal operations (joins the cluster and starts to serve regular workload).
+
+[NOTE]
+====
+As nodes in maintenance don't participate in serving usual workload, it is not recommended to execute defragmentation on several nodes at once as it reduces number of backups thus increasing the risk of partition loss.
+====
+
+When node executes defragmentation it is possible to cancel it using the following command available in control utility: *control.(sh|bat) --defragmentation cancel --host --port.*
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:43]: Please use a code block for the command here

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+====
+
+== Usage: how and when
+
+Defragmentation is costly operation in terms of disk IO so to avoid slowing down user operations it cannot be executed on regular node joined to the cluster. To execute defragmentation user needs to request it first on a particular node or set of nodes and than restart these nodes.
+
+To request defragmentation use the following command: *control.(sh|bat) --defragmentation schedule --nodes <consistentIds> [--caches <cacheNames>]*.
+
+After restart node with requested defragmentation will enter special mode called maintenance mode. Node in maintenance doesn't join the rest of the cluster but stays isolated until defragmentation is completed (or cancelled by explicit user request). After that user has to restart the node one more time: it will exit maintenance mode and returns back to normal operations (joins the cluster and starts to serve regular workload).
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:36]:
```suggestion
After the manual restart, the node with the requested defragmentation enters a special mode called maintenance mode. The node in maintenance mode does not join the rest of the cluster but remains isolated until defragmentation is completed (or canceled by explicit user request). After that, the user has to restart the node one more time: it exits maintenance mode and returns to normal operations (joining the cluster and starting to serve regular workload).
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+[NOTE]
+====
+As nodes in maintenance don't participate in serving usual workload, it is not recommended to execute defragmentation on several nodes at once as it reduces number of backups thus increasing the risk of partition loss.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:40]:
```suggestion
Nodes in maintenance mode do not participate in serving the regular workload. It is not recommended to execute defragmentation on several nodes simultaneously as it reduces the number of backups, thus increasing the risk of partition loss.
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc (hunk 8 above)
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:43]:
```suggestion

=== Stopping Defragmentation

When a node executes defragmentation, it is possible to cancel it. To stop defragmentation, run the following command available in the control utility: *control.(sh|bat) --defragmentation cancel --host --port.*
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+====
+
+When node executes defragmentation it is possible to cancel it using the following command available in control utility: *control.(sh|bat) --defragmentation cancel --host --port.*
+
+For more information about commands refer to their help.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:45]:
```suggestion

```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+[NOTE]
+====
+To reduce disk space requirements during defragmentation caches are defragmented one by one (if defragmentation of more than one cache was requested). To calculate additional space required find the cache that occupies the most disk space. The same amount of disk space is required for defragmentation at max.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:49]:
```suggestion
To reduce disk space requirements during defragmentation, caches are defragmented one by one (if a defragmentation of more than one cache is requested). To calculate additional required space, find the cache that occupies the most disk space. The same amount of disk space is required for defragmentation at max.
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+====
+
+== Conclusion
+In most situations defragmentation isn't needed as existing memory management mechanism effectively reuses memory left after data deletion. But in rare cases it may be necessary to employ it to free up disk space.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:53]:
```suggestion
In most situations, defragmentation is not necessary as existing memory management mechanism effectively reuses memory left after data deletion. But in rare cases, it may be necessary to free disk space up.
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+Its usage requires taking nodes out of normal operations so it careful planning is needed.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:55]:
```suggestion
Persistence defragmentation requires taking nodes out of their normal operations, so a careful planning is recommended.
```

**Review by Nikita-tech-writer — CHANGES_REQUESTED** [review]
@sergey-chugunov-1985 Please see my comments and the suggested changes to the PR. My point is to align this section with what is already written about the defragmentation on the GridGain documentation website.

**Comment by Mmuzaf** [comment]
@sergey-chugunov-1985 

Do we need something to resolve the comments? Can we proceed with this PR?

**Comment by Nikita-tech-writer** [comment]
@sergey-chugunov-1985  @Mmuzaf 
We should definitely apply the changes I suggested and then proceed with the PR.

File: docs/_docs/persistence/native-persistence-defragmentation.adoc (hunk 6 above)
Comment (sergey-chugunov-1985) [reply docs/_docs/persistence/native-persistence-defragmentation.adoc:32]: Disagree on the suggested change: "do not execute" implies that it is possible to do the action but is not recommended. In reality executing defragmentation on a regular node is not possible at all. How can we rephrase this to catch this meaning?

File: docs/_docs/persistence/native-persistence-defragmentation.adoc (hunk 3 above)
Comment (sergey-chugunov-1985) [reply docs/_docs/persistence/native-persistence-defragmentation.adoc:27]: I found out that historical rebalancing isn't covered with documentation yet so I removed mention of the link.

File: docs/_docs/persistence/native-persistence-defragmentation.adoc (hunk 6 above)
Comment (Nikita-tech-writer) [reply docs/_docs/persistence/native-persistence-defragmentation.adoc:32]: Then, lets leave the first version with minor changes (I've also updated the suggested change):

Defragmentation is a costly operation in terms of disk IO. **To avoid slowing down user operations, note that defragmentation cannot be executed on a regular node joined to the cluster.** To perform defragmentation, you need to request it first on a particular node or set of nodes and then restart them.

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+// limitations under the License.
+= Persistence Defragmentation
+
+== Overview
+
+Apache Ignite memory management mechanism can only create or reuse pages for user data, but it never frees them. So the files, where Ignite persists data, can only grow and never shrink.
+
+In most use cases, it does not cause any problem as once created page can be reused multiple times. However, in certain cases, it is possible that cache contains very little data but occupies large chunks of disk space because a significant volume of data was removed from the cache.
+
+Defragmentation enables user to shrink data files and claim back disk space.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:23]:
```suggestion
Defragmentation enables a user to shrink data files and claim back disk space.
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+[NOTE]
+====
+Defragmentation can only be used with historical rebalance enabled. If historical rebalancing is disabled, the server node always triggers full rebalance after the restart, which would throw away the defragmented partition. A full set of data is transferred to the node from other nodes over a network. Depending on the dataset’s size, transferring may require significant time and slow down the whole cluster as network capacity is essential to fulfill user requests.
+====
+
+== Performing Defragmentation
+
+Defragmentation is costly operation in terms of disk IO. To avoid slowing down user operations, note that defragmentation cannot be executed on a regular node joined to the cluster. To perform defragmentation, you need to request it first on a particular node or set of nodes and then restart them.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:32]:
```suggestion
Defragmentation is a costly operation in terms of disk IO. To avoid slowing down user operations, note that defragmentation cannot be executed on a regular node joined to the cluster. To perform defragmentation, you need to request it first on a particular node or set of nodes and then restart them.
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+
+=== Starting Defragmentation
+
+To request defragmentation use the following command:
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:36]:
```suggestion
To request defragmentation, use the following command:
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+control.(sh|bat) --defragmentation cancel --host --port
+----
+
+[NOTE]
+====
+To reduce disk space requirements during defragmentation, caches are defragmented one by one (if a defragmentation of more than one cache is requested). To calculate additional required space, find the cache that occupies the most disk space. The same amount of disk space is required for defragmentation at max.
+====
+
+== Conclusion
+In most situations defragmentation isn't needed as existing memory management mechanism effectively reuses memory left after data deletion. But in rare cases it may be necessary to employ it to free up disk space.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:63]:
```suggestion
In most situations, defragmentation is not necessary as existing memory management mechanism effectively reuses memory left after data deletion. But in rare cases, it may be necessary to free disk space up.
```

File: docs/_docs/persistence/native-persistence-defragmentation.adoc [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+
+Its usage requires taking nodes out of normal operations so it careful planning is needed.
```
Comment (Nikita-tech-writer) [inline docs/_docs/persistence/native-persistence-defragmentation.adoc:65]:
```suggestion
Persistence defragmentation requires taking nodes out of their normal operations, so a careful planning is recommended.
```

**Review by Nikita-tech-writer — COMMENTED** [review]
Please see my suggested changes (this time, all of them are related to wording and grmmar, not the procedure).

**Comment by sergey-chugunov-1985** [comment]
> @sergey-chugunov-1985
> 
> Do we need something to resolve the comments? Can we proceed with this PR?

@Mmuzaf I hope I'm

**Comment by sergey-chugunov-1985** [comment]
> @sergey-chugunov-1985
> 
> Do we need something to resolve the comments? Can we proceed with this PR?

@Mmuzaf I addressed (as I hope) all comments to the PR, it should be ready to merge soon.

**Review by Nikita-tech-writer — APPROVED** [review]
@sergey-chugunov-1985 Thank your for the great job!

**Comment by Nikita-tech-writer** [comment]
@Mmuzaf I suppouse this PR can be merged now.

**Comment by Mmuzaf** [comment]
@sergey-chugunov-1985 @Nikita-tech-writer 

Thank you for your efforts. I'll merge the issue when the travis checks pass.

**Review by Mmuzaf — APPROVED** [review]
